src/Modules/UngradedModules.jl

```diff
@@ -0,0 +1,3653 @@
+  free_res = free_resolution(M)[1:end-2]
+  lifted_resolution = hom(free_res, N) #TODO only three homs are neccessary
+  return homology(lifted_resolution,i)
+end
+
+#############################
+# TODO ?
+#############################
+@doc Markdown.doc"""
+    map_canonically(M::SubQuo, v::SubQuoElem)
+
+Map the element `v` to an element of the module `M` using cached 
+canonical homomorphisms between the parent Module of `v` and `M`.
+"""
+function map_canonically(M::SubQuo, v::SubQuoElem)
+  N = parent(v)
+  if N===M
+    return v
+  end
+
+  # Breadth-First Search to find path to N:
+  parent_hom = IdDict{SubQuo,ModuleMap}()
+  modules = [M]
+  found_N = false
+  for A in modules
+    for H in A.incoming_morphisms
+      B = domain(H)
+      if B!==A # on trees "B!==A" is enough!
+        if findfirst(x->x===B,modules) == nothing #if !(B in modules) doesn't work since it uses == instead of ===
+          parent_hom[B] = H
+          push!(modules,B)
+        end
+      end
+      if B===N
+        found_N = true
+        break
+      end
+    end
+    if found_N
+      break
+    end
+  end
+  if !found_N
+    throw(DomainError("There is no path of canonical homomorphisms between the modules!"))
+  end
+  result = v
+  A = N
+  while A !== M
+    H = parent_hom[A]
+    result = H(result)
+    A = codomain(H)
+  end
+  return result
+end
+
+function all_canonical_maps(M::SubQuo, N::SubQuo)
+  # from N to M
+
+  all_paths = []
+
+  function helper_dfs!(U::SubQuo, D::SubQuo, visited::Vector{<:ModuleMap}, path::Vector)
+    if U === D
+      push!(all_paths, path)
+      return
+    end
+    for neighbor_morphism in U.outgoing_morphisms
+      if findfirst(x->x===neighbor_morphism, visited) === nothing #if !(neighbor_morphism in visited) doesn't work since it uses == instead of ===
+        helper_dfs!(codomain(neighbor_morphism), D, vcat(visited, [neighbor_morphism]), union(path, [neighbor_morphism]))
+      end
+    end
+  end
+
+  helper_dfs!(N, M, Vector{ModuleMap}(), [])
+
+  morphisms = Vector{ModuleMap}()
+  for path in all_paths
+    phi = identity_map(N)
+    for h in path
+      phi = phi*h
+    end
+    push!(morphisms, phi)
+  end
+
+  return morphisms
+end
+
+#############################
+# Useful functions
+#############################
+
+@doc Markdown.doc"""
+    register_morphism!(f::ModuleMap)
+
+Cache the morphism `f` in the corresponding caches of the domain and codomain of `f`.
+"""
+function register_morphism!(f::ModuleMap)
+  push!(domain(f).outgoing_morphisms, f)
+  push!(codomain(f).incoming_morphisms, f)
+end
+
+#############################
+#TODO move to Hecke
+#  re-evaluate and use or not
+function differential(c::Hecke.ChainComplex, i::Int)
+  return map(c,length(c)-i)
+end
+
+function module_in_complex(c::Hecke.ChainComplex, i::Int)
+  return obj(c,length(c)-i)
+end
+
+getindex(c::Hecke.ChainComplex, i::Int) = module_in_complex(c,i)
+
+function Base.getindex(r::Hecke.SRow, u::UnitRange)
```

Comment:
It's a bit weird that you write `Base.getindex` in one spot and just `getindex` in the other... Also in several other places. Pick one and stick with it?